packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -106,87 +163,200 @@ export class BlobManager {
             await new Promise<void>((resolve) => this.runtime.once("attached", resolve));
         }
 
-        if (!this.runtime.connected && this.runtime.attachState === AttachState.Attached) {
-            // see https://github.com/microsoft/FluidFramework/issues/8246
-            // Avoid getting storage if we are offline since it might be undefined. In the future we will return
-            // handles immediately while offline
-            await new Promise((resolve) => this.runtime.once("connected", resolve));
-        }
+        const localId = uuid();
+        const pendingEntry: PendingBlob = {
+            blob,
+            status: PendingBlobStatus.OnlinePendingUpload,
+            handleP: new Deferred(),
+            uploadP: this.uploadBlob(localId, blob),
+        };
+        this.pendingBlobs.set(localId, pendingEntry);
 
-        const response = await PerformanceEvent.timedExecAsync(
+        return pendingEntry.handleP.promise;
+    }
+
+    private async uploadBlob(localId: string, blob: ArrayBufferLike): Promise<ICreateBlobResponse> {
+        return PerformanceEvent.timedExecAsync(
             this.logger,
             { eventName: "createBlob" },
             async () => this.getStorage().createBlob(blob),
-            { end: true, cancel: "error" },
-        );
-
-        const handle = new BlobHandle(
-            `${BlobManager.basePath}/${response.id}`,
-            this.routeContext,
-            // get() should go through BlobManager.getBlob() so handles created while detached can be redirected
-            // to the correct storage id after they are uploaded
-            async () => this.getBlob(response.id).then(async (h) => h.get()),
+            { end: true, cancel: this.runtime.connected ? "error" : "generic" },
+        ).then(
+            (response) => this.onUploadResolve(localId, response),
+            async (err) => this.onUploadReject(localId, err),
         );
+    }
 
-        if (this.runtime.attachState === AttachState.Detached) {
-            this.detachedBlobIds.add(response.id);
-            return handle;
+    private onUploadResolve(localId: string, response: ICreateBlobResponse) {
+        const entry = this.pendingBlobs.get(localId);
+        assert(!!entry, "Must have pending blob entry for uploaded blob");
+        entry.storageId = response.id;
+        if (this.runtime.connected) {
+            if (entry.status === PendingBlobStatus.OnlinePendingUpload) {
+                if (this.storageIds.has(response.id)) {
+                    entry.handleP.resolve(this.getBlobHandle(response.id));
+                    this.pendingBlobs.delete(localId);
+                } else {
+                    entry.status = PendingBlobStatus.OnlinePendingOp;
+                    this.sendBlobAttachOp(response.id);
+                }
+            } else if (entry.status === PendingBlobStatus.OfflinePendingUpload) {
+                entry.status = PendingBlobStatus.OfflinePendingOp;
```

Comment:
A similar comment explaining how an upload completes while offline would be helpful.  I'm assuming it's something like:

- upload completion is queued
- client disconnects and marks all pending uploads as offline
- client processes previous upload completion

Reply:
I added a comment but yeah, this would be too late since we already returned a handle and there may already be ops referencing it in the pending queue